Adding component to generate custom iframes into the UCSC Genome Browser

abradyIGS opened the pull request:
I did a thing. Takes a list of Ensembl IDs and creates a custom markdown/iframe sub-window for each, showing the main track view from the UCSC genome browser for the given gene.

Have integrated invocation into Titus's Snakemake wrapper. Tested, writes to dev as expected.

Dependencies:
* standard perl 5 distro
* `curl` must be available from the command line: the script currently assumes `curl` is findable via the PATH in the executing environment. if you have some other preferred way for me to make the script pull live web data, I'd be happy to update it.

Commit:
Adding component to generate custom iframes into the UCSC Genome Browser

ctb commented:
on first skim, looks great! will try running myself tomorrow. thx!

raynamharris commented:
The new feature itself looks awesome! However, I'm getting some errors. I think there is a problem with the alias gene table, because that is no longer rendering. The error message is below the screen shot: 



![Screen Shot 2022-06-29 at 12 13 46 PM](https://user-images.githubusercontent.com/7544197/176517552-a676049c-4d16-4381-9953-9468a2e748de.png)
 

```
/Users/raynamharris/anaconda3/envs/content-reg/lib/python3.10/site-packages/requests/__init__.py:109: RequestsDependencyWarning: urllib3 (1.26.9) or chardet (5.0.0)/charset_normalizer (2.0.12) doesn't match a supported version!
  warnings.warn(
2022-06-29 12:09:22,368 - WARNING - /Users/raynamharris/anaconda3/envs/content-reg/lib/python3.10/site-packages/portalocker/utils.py:210: UserWarning: timeout has no effect in blocking mode
  warnings.warn('timeout has no effect in blocking mode')

Traceback (most recent call last):
  File "/Users/raynamharris/anaconda3/envs/content-reg/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/raynamharris/anaconda3/envs/content-reg/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/raynamharris/anaconda3/envs/content-reg/lib/python3.10/site-packages/cfde_deriva/release.py", line 819, in <module>
    exit(main(*sys.argv[1:]))
  File "/Users/raynamharris/anaconda3/envs/content-reg/lib/python3.10/site-packages/cfde_deriva/release.py", line 805, in main
    release.refresh_resource_markdown()
  File "/Users/raynamharris/anaconda3/envs/content-reg/lib/python3.10/site-packages/cfde_deriva/release.py", line 396, in refresh_resource_markdown
    raise exception.StateError('Release %(id)s status=%(status)r not safe for refreshing resource markdown' % rel_row)
cfde_deriva.exception.StateError: Release 3550dad5-5bd0-4610-a9cc-18602a94f79f status='cfde_registry_rel_status:obsoleted' not safe for refreshing resource markdown

```

ctb commented:
hi @raynamharris I think I know what the problem probably is, but I don't know how it's connected to the alias table rendering? the error message above is from running `python3 -m cfde_deriva.release refresh-resources ...`, right? I have to infer that tho because I don't know what command(s) you were running to generate it.

ctb commented:
a couple second pass comments for you @abradyIGS -

first, I see your code outputs `output_pieces_gene/70-ucsc/70-ucsc_ENSG00000000003.json`, but not `output_pieces_gene/70-ucsc/70-ucsc_ENSG00000000003.md` - could you add that in so it's easy to hand-inspect the markdown without digging through JSON? (The Python code does this already, and it's been quite handy. The .md files aren't used for anything other than debugging.)

second, the dependencies on perl and curl are fine! The simplest way to handle this robustly inside of snakemake would be to use a per-rule conda environment so that perl and curl are always available when the shell command for rule `gene_json_ucsc_genome_browser_widget` is run. I'll add that in a separate PR for you to review. (This is typically what we do for any kind of library requirements that are rule-specific; it also allows version pinning which is pretty nice.)

third, I was thinking of adding this to the gallery as an example of how to write a markdown generation script in Perl. Seem ok? :)

fourth (based off the conversation from slack) it looks like you're using `curl` on _each_ Ensembl gene ID to find its location within the human genome, using the UCSC genome browser Web site. As you mentioned, this is incredibly unlikely to scale 😆  to more than a few dozen genes and is something that I'd like to avoid doing dynamically in the workflow. I'm not sure what to do about that tho - we need a way to point directly at the right genome coordinates here!

ctb commented:
can confirm this all works for me - including update and refresh. 👍

raynamharris commented:
nice! yes, that also make sense. also, the alias table is only set to build for 1 or 2 genes, so that's probably why it wasn't there.

abradyIGS commented:
* will add md outputs; seemed like they didn't connect to anything downstream so i left them out, but no sweat to add
* cool on rule addition, will await PR
* gallery addition is fine by me, i'll add a brief header to the code when i add the markdown dump
* the target genome browser URLs aren't just chromosome-coordinate-dependent (although i might be able to generate some as an alternative); there's some extra resolution & cross-referencing that i'm relying on their server to do when i ask for info about an Ensembl ID in the way that I'm asking for it. this process surely will not scale in its present form; they do have an API, but looking at its available endpoints, it seems there's a mountain of logic and information available via the web UI that UCSC doesn't make accessible via the API. still, i'll need to find some way to get this data, either live in bulk or offline. will do.

note: all of the above will happen tomorrow, as i am about to leave for new jersey. back late tonight.

abradyIGS commented:
Am working on all of above right now: i think i've found a way to generate links without pinging UCSC at all; stand by

Commit:
Added .md outputs; removed dependency on curl; added provenance header to script; rewrote URL-generation to avoid any live queries to UCSC

abradyIGS commented:
All changes complete and have tested okay; renewing review request (if there's a button for that, LMK and i will press it).

* added .md file generation
* `curl` no longer used: I wrote a new URL generator that uses preloaded coordinate data instead of live queries to UCSC
* added provenance header to script, send to gallery when ready

ctb commented on scripts/build-markdown-pieces-ucsc-genome-browser-widget.pl:
```diff
@@ -0,0 +1,173 @@
+# 
+# Chaise markdown guide: https://github.com/informatics-isi-edu/ermrestjs/blob/master/docs/user-docs/markdown-formatting.md
+# 
+# Creation date: 2022-06-29
```
as a side note, this info is kept automatically in the git history, too :)

abradyIGS replied:
cool, another piece of insight for my extremely belated transition to multi-person source control. thank 👍

ctb replied:
(it's in single person source control too 😜 )

ctb reviewed (approved):
thanks!

ctb committed:
Merge branch 'main' into add_new_plugins

ctb commented:
inquiry - do we want to enable these on all available genes, or all genes with biosamples, or...?

abradyIGS commented:
i think all available genes makes sense; if there's a gene in our portal DB, it'll be connected either to a biosample or to a collection, and so should be findable.

raynamharris commented:
i agree, seems good to annotate all the genes in portal, but we must check they they are in the validation ensemble_genes.tsv file. see https://github.com/nih-cfde/update-content-registry/issues/22

abradyIGS commented:
minor correction -- file to check is `data/inputs/DEV_PORTAL__available_genes__2022-07-01.txt` -- the `ensembl_genes.tsv` file contains all of Ensembl, and in particular includes gene IDs [available to the DCCs but] not available in the portal [because nobody used them]. see #21

abradyIGS commented:
(in fact if you just repoint the snakemake invocation of the UCSC iframe builder script to `DEV_PORTAL__available_genes__2022-07-01.txt` instead of `gene_IDs_for_UCSC_genome_browser_widget.txt` (the script's currently-designated list of input genes), it should do what you want without further edits (alternatively you could replace the latter file with a symlink pointing to the former file, instead of embedding a filename that will change with future updates into the middle of the makefile, but i dunno how well git copes with symlinks). I'd start a new PR and just do it, but (a) i'm wrapping up for the day and figure this is straightforward enough to leave to others if you want it done before i get back, and (b) you might not want things done exactly this way, so i'm leaving it alone by default.)
